cou/steps/plan.py

```diff
@@ -0,0 +1,58 @@
+# Copyright 2023 Canonical Limited.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Upgrade planning utilities."""
+
+import logging
+import sys
+from argparse import Namespace
+from typing import Any
+
+from cou.steps import UpgradeStep
+from cou.steps.backup import backup
+
+
+def generate_plan(args: Namespace) -> UpgradeStep:
+    """Generate plan for upgrade."""
+    logging.info(args)  # for placeholder
+    plan = UpgradeStep(description="Top level plan", parallel=False, function=None)
+    plan.add_step(
+        UpgradeStep(description="backup mysql databases", parallel=False, function=backup)
+    )
+    return plan
+
+
+def apply_plan(upgrade_plan: Any) -> None:
+    """Apply the plan for upgrade."""
+    result = input(upgrade_plan.description + "[Continue/abort/skip]")
+    if result.casefold() == "c".casefold():
+        if upgrade_plan.function is not None:
+            if upgrade_plan.params:
+                upgrade_plan.function(upgrade_plan.params)
+            else:
+                upgrade_plan.function()
```

Comment:
What I understand from your comment is "Should we inherit all steps from UpgradeStep"?

Reply:
No I mean that here we can run `upgrade_plan.run()` and this will do `self.function(self.params)` in background + all logging, if conditions, etc.

Reply:
Running logic will be more complicated this is simple placeholder. Because it will also include parallel substep runs also. Thus I would like to separate it from the UpgradeStep. However we can start in the UpgradeStep and seperate it if needed.

Reply:
Running logic yes, but you will always run `upgrade_plan.function()` or `upgrade_plan.function(upgrade_plan.params)`, so why do not make step executable like this:
```python
class UpgradeStep:
    def __init__(
        self,
        description: str,
        parallel: bool,
        function: Optional[Callable],
        **params: Optional[Any],
    ):
        """Initialize upgrade step."""
        self.parallel = parallel
        self.description = description
        self.sub_steps: List[UpgradeStep] = list[UpgradeStep]()
        self.params = params or {}
        self.function = function

    def __call__(self):
        if self.function is not None:
            self.function(**self.params)
```

Reply:
Maybe it's not good idea, but when I saw it I though that step should be self alone which could be run without accessing params or functions.... I mean that `upgrade_plan()` seam better to me than accessing function and params (e.g. `upgrade_plan.params` and `upgrade_plan.function`), even better would be to have function and params as private (e.g. `upgrade_plan._func` and `upgrade_plan._params`)

Reply:
moved to UpgradeStep